File: ipa_canopen_core/scripts/os_shell.py
#!/usr/bin/env python
from can import *
from canutils import *
import sys
import time
import itertools

def send_segmented_sdo(port, can_id, data, toggle, more_segments):
    obj_id = can_id + 0x600 

    data_len = len(data)
    if data_len > 7:
      print("Error: data was > 7 bytes long")
      sys.exit(-1)

    toggle_bit = 0b10000 if toggle else 0
    more_segments_bit = 0b1 if more_segments else 0
    size_bits = (7-data_len) << 1
    first_byte = [ 0x01 + toggle_bit + more_segments_bit + size_bits]
    packet_data = itertools.chain(first_byte, data)
    port.send_msg(obj_id, list(packet_data), 0)

def get_sdo_key_from_data(data):
    return SDOKey(data[1] + (data[2] << 8), data[3])

def chunks(l, n):
    """ Yield successive n-sized chunks from l. """
    for i in xrange(0, len(l), n):
            yield l[i:i+n]

def send_segmented_data(port, can_id, sdo_key, data):
    port.init_send_segmented_sdo(can_id, sdo_key)
    sleepms(1000)
    res_data = port.recv_msg().data
    received_sdo_key = get_sdo_key_from_data(res_data)
    if received_sdo_key != sdo_key:
      print "ERROR: unexpected response while init segmented sdo send"
    splitted_data = list(chunks(data, 7))

    toggle = False
    chunk_num = 1

    for chunk in splitted_data:
      more_segments = True
      if len(splitted_data) == chunk_num:
        more_segments = False 
      send_segmented_sdo(port, can_id, chunk, toggle, more_segments)
      sleepms(1000)
      res_data = port.recv_msg().data
      expected_first_byte = 0x20
      toggle_bit = 0b10000 if toggle else 0
      if res_data[0] != (0x20 + toggle_bit):
        print("ERROR bad server response ")
      toggle = not toggle
      chunk_num = chunk_num + 1
      
def str_to_data_g(string):
    for c in string: 
      yield ord(c)

def str_to_data(string):
    return list(str_to_data_g(string))

def init_os_mode(p, can_id):
    sdo_key = SDOKey(0x1024, 0)
    p.send_sdo(can_id, sdo_key, 0, True)
    sleepms(3000)
    resp = p.recv_msg()
    expectedResData = [0x60, 0x24, 0x10, 0, 0, 0, 0, 0]
    if not same_data(expectedResData, resp.data):
        raise Exception("unable to set evaluate immediately mode")
    sleepms(100)
    p.clean_buffers()
    sdo_key = SDOKey(0x1024, 0)
    p.send_sdo(can_id, sdo_key, 0, True)
    sleepms(3000)

def start_shell(can_id):
    p = CanPort(0, 0)
    p.open()
    p.config_port(mode.BIT11, 0, 0x7EE, baud_rate.CAN_BAUD_500K) # TODO: check all parameters
    sleepms(100)
    p.clean_buffers()

    p.reset_communication()
    sleepms(2000)
    #p.start_remote_node(can_id)
    #sleepms(4000)

    p.clean_buffers()

    sdo_key = SDOKey(0x1024, 0)
    p.send_sdo(can_id, sdo_key, 0, True)

    sleepms(3000)
    resp = p.recv_msg()
    expectedResData = [0x60, 0x24, 0x10, 0, 0, 0, 0, 0]
    if not same_data(expectedResData, resp.data):
        raise Exception("unable to set evaluate immediately mode")

    sleepms(100)
    p.clean_buffers()
    command_sdo_key = SDOKey(0x1023, 1)

    
    stop = False
    print ("enter 'q' or 'exit' to finish the program")
    while not stop:
      s = raw_input("-->:")
      if s == "q" or s == "exit":
        stop = True
      else:
        send_segmented_data(p, can_id, command_sdo_key, str_to_data(s))
        sleepms(400)


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("usage: set-canid CAN_ID")
        sys.exit(-1)
    can_id = int(sys.argv[1])
    start_shell(can_id)

File: ipa_canopen_core/src/os_shell.cpp
#include "ipa_canopen_core/can_port.h"

#include <iostream>
#include <time.h>
#include <unistd.h>
#include "assert.h"
#include <math.h>
#include <string>

using namespace std;

SDOkey OS_COMMAND_MODE(0x1024, 0);
SDOkey OS_COMMAND_PROMPT(0x1023, 0);

void mssleep(int ms) {
  usleep(ms * 10e3);
}

SDOkey getSdoKeyFromPacket(const CAN_PACKET& msg) {
  return SDOkey( msg.data[1] + (msg.data[2] << 8), msg.data[3]);
}

void sendSegmentedSdo(CanPort& p, int canId, std::vector<BYTE> data, bool toggle, bool moreSegments) {
  int objId = canId + 0x600;

  BYTE firstByte = 0x01;
  if(toggle) {
    firstByte += 0b10000;
  }
  if(moreSegments) {
    firstByte += 0b1;
  }
  firstByte += (7-data.size()) << 1;
  data.insert(data.begin(), firstByte);
  p.sendMsg(objId, data, 0);
}

void sendSegmentedData(CanPort& p, int canId, SDOkey sdoKey, std::vector<BYTE> data)
{
  p.initSendSegmentedSdo(canId, sdoKey);
  mssleep(1000);
  SDOkey receivedSdoKey = getSdoKeyFromPacket(p.rcvMsg());

  if(receivedSdoKey != sdoKey) {
    std::cerr << "sdo key was wrong during send segmented data" << std::endl;
    exit(-1);
  }

  bool toggle = false;
  int chunk_num = ceil(data.size()/7.0f);
  for(int cur_chunk = 0; cur_chunk < chunk_num; cur_chunk++) {
    bool moreSegments = cur_chunk+1 < chunk_num;
    std::vector<BYTE> chunk;
    for(int i = cur_chunk*7; i < min( (cur_chunk+1) * 7, (int) data.size()); i++) {
      chunk.push_back(data[i]);
    }
    sendSegmentedSdo(p, canId, chunk, toggle, moreSegments);
    CAN_PACKET res = p.rcvMsg();
    BYTE expectedFirstByte = 0x20;
    if(toggle) {
      expectedFirstByte += 0b10000;
    }
    if (res.data[0] != expectedFirstByte) {
      cerr << "ERROR: bad server response" << endl;
    }
    toggle = !toggle;
  }

}

std::vector<BYTE> dataToVector(const CAN_PACKET& msg) {
  std::vector<BYTE> v;
  for(int i = 0; i < msg.len; i++) {
    v.push_back(msg.data[i]);
  }
  return v;
}

bool areEquals(const std::vector<BYTE>& l, const std::vector<BYTE>& r) {
  return std::equal(l.begin(), l.end(), r.begin());
}

bool assertSameData(const CAN_PACKET& msg, const std::vector<BYTE>& data) {
  std::vector<BYTE> l = dataToVector(msg);
  if(!areEquals(l, data)) {
    throw new CanException("Different data");
  }
}


std::vector<BYTE> stringToVec(const std::string& s) {
  std::vector<BYTE> vec;

  for(int i = 0; i < s.size(); i++) {
    vec.push_back(s[i]);
  }

  return vec;
}

void startShell(int canId) {
  cout << " starting shell for canId: " << canId << endl;

  CanPort p = CanPort(0, 0);
  p.open();

  p.configPort(BIT11, 0, 0x7EE, CAN_BAUD_500K);
  p.cleanBuffers();

  mssleep(100);
  p.sendSDO(canId, OS_COMMAND_MODE, 0);

  mssleep(1000);
  CAN_PACKET res = p.rcvMsg();
  assertSameData(res, {0x60, 0x24, 0x10, 0,   0, 0, 0, 0});

  mssleep(500);
  p.cleanBuffers();


  while(true) {
    string s;
    cout << "-->";
    cin >> s;
    cout << endl;
    if(s == "exit" || s == "q") {
      break;
    }
    std::vector<BYTE> data = stringToVec(s);
    sendSegmentedData(p, canId, OS_COMMAND_PROMPT, data);
  }
}

int main(int argc, char** argv) {
  if (argc != 2) {
    cerr << "usage: " << argv[0]  << " canId " << endl;
  }

  int canId = atoi(argv[1]);
}
